src/claim/Claim.utils.ts

```diff
@@ -51,52 +65,129 @@ export function toJsonLD(
+  // apply defaults
+  const defaults = {
+    hasher: defaultHasher,
+  }
+  const { hasher, nonces, nonceGenerator } = { ...defaults, ...options }
+  // getNonce <- if nonces map then take from map, else use nonceGenerator function
+  // if no nonceGenerator given, unsalted hashes are returned
+  const getNonce: HashingOptions['nonceGenerator'] =
+    typeof nonces === 'object'
+      ? (key) => {
+          const nonce = nonces[key]
+          // nonce map must be exhaustive
+          // TODO: use sdk error module
+          if (!nonce)
+            throw new Error(
+              `no nonce given for statement with nonce key ${key}`
+            )
+          return nonce
+        }
+      : nonceGenerator
+  // map over statements, use key to retrieve nonce, then hash statement & nonce
+  return Object.entries(statements).map(([key, statement]) => {
+    const nonce = getNonce ? getNonce(key) : undefined
     return {
       key,
       nonce,
-      // to simplify validation, the salted hash is computed over unsalted hash (nonce key) & nonce
-      hash: hasher(key, nonce),
+      hash: hasher(statement, nonce),
     }
   })
 }
 
+/**
+ * Produces salted hashes of individual statements comprising a (partial) [[IClaim]] to enable selective disclosure of contents. Can also be used to reproduce hashes for the purpose of validation.
+ *
+ * @param claim Full or partial [[IClaim]] to produce statement hashes from.
+ * @param options Object containing optional parameters.
+ * @param options.canonicalisation Canonicalisation routine that produces an array of statement strings from the [IClaim]. Default produces individual `{"key":"value"}` JSON representations where keys are transformed to expanded JSON-LD.
+ * @param options.nonces Optional map of nonces as produced by this function.
+ * @param options.nonceGenerator Nonce generator as defined by [[hashStatements]] to be used if no `nonces` are given. Default produces random UUIDs (v4).
+ * @param options.hasher The hasher to be used. Required but defaults to 256 bit blake2 over `${nonce}${statement}`.
+ * @returns An array of salted `hashes` and a `nonceMap` where keys correspond to unsalted statement hashes.
+ */
+export function hashClaimContents(
+  claim: PartialClaim,
+  options: HashingOptions & {
+    canonicalisation?: (claim: PartialClaim) => string[]
+  } = {}
+): { hashes: string[]; nonceMap: Record<string, string> } {
+  // apply defaults
+  const defaults = {
+    hasher: defaultHasher,
+    nonceGenerator: () => uuid(),
+    canonicalisation: makeStatementsJsonLD,
+  }
+  const { hasher, nonces, nonceGenerator, canonicalisation } = {
+    ...defaults,
+    ...options,
+  }
+  // use canonicalisation algorithm to make hashable statement strings
+  const statements = canonicalisation(claim)
+  // generate unsalted hashes from statements as a first step
+  const unsalted = hashStatements(statements, {
+    hasher,
+    nonceGenerator: undefined,
```

Comment:
This is the same as just leaving it out. Why did you add it explicitly?

Reply:
in case we would set a default generator, this would force it to undefined (=no nonces)

Reply:
ah ok